src/frontends/src/query_frontend.hpp

```diff
@@ -318,25 +318,35 @@ class RequestHandler {
           }
         }
 
-        folly::Future<Response> prediction = decode_and_handle_predict(
-            request->content.string(), name, versioned_models, policy,
-            latency_slo_micros, input_type);
-
-        prediction
-            .then([response, app_metrics](Response r) {
-              // Update metrics
-              if (r.output_is_default_) {
-                app_metrics.default_pred_ratio_->increment(1, 1);
-              } else {
-                app_metrics.default_pred_ratio_->increment(0, 1);
+        folly::Future<std::vector<folly::Try<Response>>> predictions =
+            decode_and_handle_predict(request->content.string(), name,
+                                      versioned_models, policy,
+                                      latency_slo_micros, input_type);
+
+        predictions
+            .then([response,
+                   app_metrics](std::vector<folly::Try<Response>> tries) {
+              std::string final_content;
+              for (auto t : tries) {
+                try {
+                  Response r = t.value();
+                  if (r.output_is_default_) {
+                    app_metrics.default_pred_ratio_->increment(1, 1);
+                  } else {
+                    app_metrics.default_pred_ratio_->increment(0, 1);
+                  }
+                  app_metrics.latency_->insert(r.duration_micros_);
+                  app_metrics.num_predictions_->increment(1);
+                  app_metrics.throughput_->mark(1);
+
+                  std::string content = get_prediction_response_content(r);
+                  final_content += content + "\n";
```

Comment:
This will turn into a stringstream appends (e.g. `final_content << content << "\n";`)